src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -81,45 +76,84 @@ class ExportEntitiesByTypeActor(rawlsDAO: RawlsDAO,
   def ExportEntities = streamEntities()
 
   /**
-    * Two basic code paths
-    *
-    * For Collection types, write the content to temp files, zip and return.
+    * Calls entitiesToTempFile() to write entity TSV to a temp file,
+    * then creates an HttpResponse containing that temp file's contents as an attachment.
     *
-    * For Singular types, pipe the content from `Source` -> `Flow` -> `Sink`
-    * Source generates the entity queries
-    * Flow executes the queries and sends formatted content to chunked response handler
-    * Sink finishes the execution pipeline
-    *
-    * Handle exceptions directly by completing the request.
+    * @see [[entitiesToTempFile()]]
     */
   def streamEntities(): Future[HttpResponse] = {
     val keepAlive = Connection("Keep-Alive")
 
+    entitiesToTempFile() map { tempFile =>
+      val (contentType, fileName) = if (modelSchema.isCollectionType(entityType)) {
+        (ContentTypes.`application/octet-stream`, s"$entityType.zip")
+      } else {
+        (ContentType(MediaTypes.`text/tab-separated-values`, HttpCharsets.`UTF-8`), s"$entityType.tsv")
+      }
+      val disposition = `Content-Disposition`.apply(ContentDispositionTypes.attachment, Map("filename" -> fileName))
+      HttpResponse(entity = HttpEntity.fromFile(contentType, tempFile.toJava), headers = List(keepAlive, disposition))
+    }
+  }.recoverWith {
+    // Standard exceptions have to be handled as a completed request
+    case t: Throwable => handleStandardException(t)
+  }
+
+  /**
+    * Calls entitiesToTempFile() to write entity TSV to a temp file,
+    * then uploads from that temp file to the workspace's bucket.
+    *
+    * @see [[entitiesToTempFile()]]
+    *
+    */
+  def streamEntitiesToWorkspaceBucket(): Future[GcsPath] = {
+    // retrieve workspace so we can get its bucket
+    rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)(userInfo) flatMap { workspaceResponse =>
+      val workspaceBucket = GcsBucketName(workspaceResponse.workspace.bucketName)
+
+      // verify write permissions
+      val isPermitted = workspaceResponse.accessLevel match {
+        case Some(accessLevel) => accessLevel >= WorkspaceAccessLevels.Write
+        case None => false
+      }
+      if (!isPermitted)
+        throw new FireCloudExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Forbidden, s"You must have at least write access."))
```

Comment:
this permission implementation - checking what the workspace reports instead of querying Sam - is strange but it 1) saves an extra API call and 2) is consistent with how Orch checks permissions other places. I waffle on whether it is the right thing to do. But in the end, Orch isn't the permission enforcer; this check just exists to give a nicer error message in cases where the user only has read permissions.